File: src/main/java/it/unibo/crabinv/View/MainMenu.java
package it.unibo.crabinv.View;

import it.unibo.crabinv.Controller.audio.AudioController;
import it.unibo.crabinv.Controller.i18n.LocalizationController;
import it.unibo.crabinv.Model.audio.SFXTracks;
import it.unibo.crabinv.Model.i18n.TextKeys;
import it.unibo.crabinv.SceneManager;
import javafx.application.Platform;
import javafx.geometry.Pos;
import javafx.scene.layout.Pane;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.scene.control.Button;
import javafx.scene.control.Label;


public class MainMenu {
    private final SceneManager sceneManager;
    private final LocalizationController loc;
    private final AudioController audio;

    public MainMenu(SceneManager sceneManager, LocalizationController loc, AudioController audio) {
        this.sceneManager = sceneManager;
        this.loc = loc;
        this.audio = audio;
    }

    public Pane getView() {
        Pane pane = new StackPane();
        VBox mainColumn = new VBox(30);
        mainColumn.setAlignment(Pos.CENTER);


        Label title = new Label("Crab Invaders");
        title.getStyleClass().add("menu-title");

        mainColumn.getChildren().addAll(
                title,
                createMenuButton(TextKeys.PLAY, () -> System.out.println("PLAY")),
                createMenuButton(TextKeys.SHOP, () -> System.out.println("SHOP")),
                createMenuButton(TextKeys.RUN_LOG, () -> System.out.println("RUN_LOG")),
                createMenuButton(TextKeys.SETTINGS, sceneManager::showSettings),
                createMenuButton(TextKeys.EXIT_GAME, Platform::exit)
        );
        pane.getChildren().add(mainColumn);
        return pane;
    }

    private Button createMenuButton(TextKeys key, Runnable action) {
        Button menuButton = new Button(loc.getString(key));
        menuButton.getStyleClass().add("menu-button");

        menuButton.setMinWidth(220);

        menuButton.focusedProperty().addListener((_, _, newValue) -> {
            if (newValue) {
                audio.playSFX(SFXTracks.MENU_HOVER);
            }
        });

        menuButton.setOnAction(_ -> {
            audio.playSFX(SFXTracks.MENU_SELECT);
            action.run();
        });

        return menuButton;
    }


}

File: src/main/java/it/unibo/crabinv/view/MainMenu.kt
package it.unibo.crabinv.view

import it.unibo.crabinv.SceneManager
import it.unibo.crabinv.controller.core.audio.AudioController
import it.unibo.crabinv.controller.core.i18n.LocalizationController
import it.unibo.crabinv.model.core.audio.SFXTracks
import it.unibo.crabinv.model.core.i18n.TextKeys
import javafx.application.Platform
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox

/**
 * This is the view of the Main Menu.
 * Part of this was adapted from LLM by Mosè Barbieri
 * @param sceneManager it's the manager of the scenes, it moves them when needed
 * @param loc the needed manager for the translation
 * @param audio the audio manager needed for the sounds of the buttons
 */
class MainMenu(
    private val sceneManager: SceneManager,
    private val loc: LocalizationController,
    private val audio: AudioController,
) {
    val view: Pane
        /**
         * it's the getter of the view of the menu.
         * @return the view of the menu
         */
        get() {
            val pane: Pane = StackPane()
            val mainColumn =
                VBox(ViewParameters.DEFAULT_SPACING_MAINMENU.toDouble())
            mainColumn.alignment = Pos.CENTER

            val title = Label("Crab Invaders")
            title.styleClass.add("menu-title")

            mainColumn.children.addAll(
                title,
                createMenuButton(TextKeys.PLAY) { sceneManager.showGame() },
                createMenuButton(TextKeys.SHOP) { sceneManager.showShop() },
                createMenuButton(TextKeys.RUN_LOG) { sceneManager.showMemorial() },
                createMenuButton(TextKeys.SETTINGS) { sceneManager.showSettings() },
                createMenuButton(TextKeys.EXIT_GAME) { Platform.exit() },
            )
            pane.children.add(mainColumn)
            return pane
        }

    private fun createMenuButton(
        key: TextKeys,
        action: Runnable,
    ): Button {
        val menuButton = Button(loc.getString(key))
        menuButton.styleClass.add("app-button")

        menuButton
            .focusedProperty()
            .addListener(
                ChangeListener { _: ObservableValue<out Boolean?>?, _: Boolean?, newValue: Boolean? ->
                    if (newValue == true) {
                        audio.playSFX(SFXTracks.MENU_HOVER)
                    }
                },
            )
        menuButton
            .onAction =
            EventHandler { _: ActionEvent? ->
                audio.playSFX(SFXTracks.MENU_SELECT)
                action.run()
            }

        return menuButton
    }
}
